agent/tool-scripts/README.md

```diff
@@ -0,0 +1,219 @@
+differences where appropriate.
+
+The `external-data-source`, `node-exporter`, and `dcgm` "tools" are handled
+separately since they operate a bit different from the rest.  See below for
+more details.
+
+## The Tool Life Cycle
+
+There are 4 phases, if you will, that the Tool Meister sub-system offers. The
+first phase is for executing all the "persistent" tools.  Once the Tool
+Meister sub-system is up and running, this phase is started first (required),
+and is ended before the Tool Meister sub-system is shut down (required).  The
+`pbench-init-tools` CLI interface begins the "persistent" tools phase, and the
+`pbench-end-tools` closes out the phase.  More about "persistent" tools below.
+
+The second phase is for executing all the "transient" tools (described below).
+This phase must be started after the "persistent" tools phase begins, and must
+end before the "persistent" tools phase ends.  However, the phase can repeated
+as many times as the caller likes.  The `pbench-start-tools` CLI begins the
+phase by start all the "transient" tools.  Then the caller must invoke the
+`pbench-stop-tools` CLI to shut down all the "transient" tools, ending this
+phase.
+
+The third phase is for all transient tools to send their data back to the host
+driving all the local and remote Tool Meisters.  This phase can happen after
+all "transient" tool phases have executed, but must occur before ending the
+"persistent" tools phase.  The execution of the `pbench-send-tools` CLI
+interface encapsulates this phase.
+
+The fourth, and final, phase is the post-processing for "transient" tool data.
+All the post-processing for a given "transient" tool phase is invoked locally
+on the host driving all the Tool Meisters.  The execution of the
+`pbench-postprocess-tools` CLI interface handles this phase for a given
+"transient" tool phase, and must be invoked after the `pbench-send-tools` for
+a given phase.
+
+Inspecting the `agent/bench-scripts` directory for how the various scripts
+invoke the aforementioned CLIs will give you a sense of how these phases are
+used.
+
+## Examples of the Four Phases
+
+The standard way tools are invoked:
+
+```
+  $ pbench-init-tools
+  $ pbench-start-tools
+  $ pbench-stop-tools
+  $ pbench-send-tools
+  $ pbench-postprocess-tools
+  $ pbench-end-tools
+```
+
+A benchmark might require multiple transient tool phases:
+
+```
+  $ pbench-init-tools
+  $
+  $ pbench-start-tools
+  $ pbench-stop-tools
+  $ pbench-send-tools
+  $ pbench-postprocess-tools
+  $
+  $ pbench-start-tools
+  $ pbench-stop-tools
+  $ pbench-send-tools
+  $ pbench-postprocess-tools
+  $
+  $ pbench-end-tools
+```
+
+Another benchmark environment might need to delay sending data and
+post-processing at a later time:
+
+```
+  $ pbench-init-tools
+  $
+  $ pbench-start-tools
+  $ pbench-stop-tools
+  $
+  $ pbench-start-tools
+  $ pbench-stop-tools
+  $
+  $ # All workloads done, its safe to send data remotely
+  $
+  $ pbench-send-tools
```

Comment:
Do we need two `pbench-send-tools`?

Reply:
Yes, we send only the tool data for a specific start/top pair.